runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Timer.java

```diff
@@ -40,92 +52,201 @@
-  public static <T> Timer<T> of(Instant timestamp, @Nullable T payload) {
-    return new AutoValue_Timer(timestamp, payload);
+  /** Returns a cleared timer for the given userKey, dynamicTimerTag and windows. */
+  public static <T> Timer<T> cleared(
+      T userKey, String dynamicTimerTag, Collection<? extends BoundedWindow> windows) {
+    return new AutoValue_Timer(userKey, dynamicTimerTag, windows, true, null, null, null);
   }
 
+  /** Returns the key that the timer is set on. */
+  public abstract T getUserKey();
+
+  /**
+   * Returns the tag that the timer is set on. This tag is {@code ""} when the timer is for a {@link
+   * TimerSpec}.
+   */
+  public abstract String getDynamicTimerTag();
+
+  /** Returns the windows which is associated with the timer. */
+  public abstract Collection<? extends BoundedWindow> getWindows();
+
+  /** Returns whether the timer is going to be cleared. */
+  public abstract boolean getClearBit();
+
   /**
-   * Returns the timestamp of when the timer is scheduled to fire.
+   * Returns the timestamp of when the timer is scheduled to fire. This field is nullable only when
+   * the timer is being cleared.
    *
-   * <p>The time is relative to the time domain defined in the {@link
+   * <p>The time is absolute to the time domain defined in the {@link
    * org.apache.beam.model.pipeline.v1.RunnerApi.TimerFamilySpec} that is associated with this
    * timer.
    */
-  public abstract Instant getTimestamp();
+  @Nullable
+  public abstract Instant getFireTimestamp();
 
-  /** A user supplied payload. */
+  /**
+   * Returns the watermark that the timer is supposed to be held. This field is nullable only when
+   * the timer is being cleared.
+   */
   @Nullable
-  public abstract T getPayload();
+  public abstract Instant getHoldTimestamp();
+
+  /**
+   * Returns the paneinfo that is related to the timer. This field is nullable only when the
+   * timer is being cleared.
+   */
+  @Nullable
+  public abstract PaneInfo getPane();
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof Timer)) {
+      return false;
+    }
+    Timer<?> that = (Timer<?>) other;
+    if (this.getClearBit()) {
+      return Objects.equals(this.getUserKey(), that.getUserKey())
+          && this.getDynamicTimerTag().equals(that.getDynamicTimerTag())
+          && (this.getClearBit() == that.getClearBit())
+          && Objects.equals(this.getWindows(), that.getWindows());
+    }
+    return Objects.equals(this.getUserKey(), that.getUserKey())
+        && this.getDynamicTimerTag().equals(that.getDynamicTimerTag())
+        && (this.getClearBit() == that.getClearBit())
+        && this.getFireTimestamp().equals(that.getFireTimestamp())
+        && this.getHoldTimestamp().equals(that.getHoldTimestamp())
+        && Objects.equals(this.getWindows(), that.getWindows())
+        && Objects.equals(this.getPane(), that.getPane());
+  }
```

Comment:
This will make the method less error prone in case the contract changes.
```suggestion
    return Objects.equals(this.getUserKey(), that.getUserKey())
        && Objects.equals(this.getDynamicTimerTag(), that.getDynamicTimerTag())
        && Objects.equals(this.getWindows(), that.getWindows())
        && (this.getClearBit() == that.getClearBit())
        && Objects.equals(this.getFireTimestamp(), that.getFireTimestamp())
        && Objects.equals(this.getHoldTimestamp(), that.getHoldTimestamp())
        && Objects.equals(this.getPane(), that.getPane());
  }
```